### ISSUE TITLE
how to advertise server limits

### ISSUE THREAD

**Issue by pomakis** [opengeospatial/OGC-API-Sprint-August-2020#10]
As discussed on https://gitter.im/opengeospatial/OGC-API-Sprint-August-2020 on 2020-08-07, a server needs to be able to advertise certain server limits such as the maximum number of coverage samples a client is allowed to request.  There are similar server limits in "OGC API - Maps", such as maxWidth and maxHeight.  The "OGC API - Maps" limits are advertised in the maps metadata document.  But I'm not sure where "OGC API - Coverages" limits should go.

A few possible options:
- in /collections/{collectionId}/coverage/metadata
- in /collections/{collectionId}/coverage/limits (a new endpoint)
- in /collections/{collectionId} (in a common "limits" object)
- in /limits, a common endpoint for all OGC API specification to throw their own
 content into

**Comment by jerstlouis** [comment]
@pomakis my would would be in /limits
Definitely not /metadata as this is specifically the CIS metadata content of that coverage.

**Comment by cmheazel** [comment]
If it is a limit imposed by the server then it can go on the landing page.
If it is a limit on the individual coverage, then it can go in the collection Information object.
These would be extensions to the standard schemas.

**Comment by joanma747** [comment]
This seems an idea for a CIS extension. For the moment the CIS is agnostic of the "server" capacity and only describes the coverage.  However I see no reason to extend it in this direction. 

If we imagine a "limits" as part of the CIS description at the same level of domainSet, rangeset etc..., then, it automatically becomes available as /collections/{collectionId}/coverage/limits as one of the @pomakis suggestions at the beginning of this issue.

**Comment by jerstlouis** [comment]
@joanma747 I think what we are thinking here is outside of the CIS data model. The exact same coverage may be offered on multiple servers with different limits.

**Comment by joanma747** [comment]
>The exact same coverage may be offered on multiple servers with different limits.
Agreed in principle. That does not preclude that the server uses the CIS model to communicate it in an extra class. We could start discussing if the fact that the data is too big to be served in a single piece is the server's fault or the data's fault. :-)

In any case, to me, it is something at the same level than rangetype, domainset etc even it is not part of the CIS. I'm reluctant to expose particularities of coverages in the landingpage or in /collections/collectionid so I cannot see any other place than "below" /collections/collectionId/coverage. 

Just to contradict myself, it could be in /collections/collectionid if it is expressed "neutral" to the data model. e.g, expressed as bbox-limits in projected coordinates.

**Comment by joanma747** [comment]
>The exact same coverage may be offered on multiple servers with different limits.
Agreed in principle. That does not preclude that the server uses the CIS model to communicate it in an extra class. We could start discussing if the fact that the data is too big to be served in a single piece is the server's fault or the data's fault. :-)

In any case, to me, it is something at the same level than rangetype, domainset etc even it is not part of the CIS. I'm reluctant to expose particularities of coverages in the landingpage or in /collections/collectionid so I cannot see any other place than "below" /collections/collectionId/coverage. 

Just to contradict myself, it could be in /collections/collectionid if it is expressed "neutral" to the data model. e.g, expressed as bbox-limits in projected coordinates. E.g. bbox-limits: [20000,20000] says that you cannot ask for an area bigger than 20000*20000 meters

**Comment by jerstlouis** [comment]
@joanma747 bbox in projected coordinates doesn't work if the server allows serving that full BBOX as a scaled down version :)

As originally posted the suggestion was not data-specific, and `/limits` specifying things expressed differently for different APIs make perfect sense to me...  or even if it were `/limits/coverage` and `/limits/features`.

I'm not sure a coverage-specific limit has a practical use case?

**Comment by joanma747** [comment]
> @joanma747 bbox in projected coordinates doesn't work if the server allows serving that full BBOX as a scaled down version :)

:-). Agreed. bboxLimits in projected coordinates should be accompanied by a scaleLimits that will tell us the minimum scale denominator supported.

**Comment by cmheazel** [comment]
API-Coverage inherits the `limits` parameter from API-Common. The purpose of this parameter is to limit the size of a response and to support retrieval of the remaining data. It also allows for a default limit which is imposed by the server. What isn't defined is the units of measure. In API-Features it is a count of the Features returned. I suggest bytes for API-Coverages.

**Comment by jratike80** [comment]
If you compress the output data you know the number of bytes a bit too late IMHO.

**Comment by jerstlouis** [comment]
@cmheazel but the server, or even the colleciton, should support both features and coverages so you would need separate limits.

**Comment by cmheazel** [comment]
@jratike80 If you compress the output then I would use the uncompressed size. If you truncate a compressed stream the client may not be able to decompress it.

**Comment by cmheazel** [comment]
@jerstlouis  The parameter is the same for any resource type, but the effect of the parameter is resource specific. So a limit of 10 returns:
1) 10 Features
2) 10 bytes (or whatever we decide) of coverage
3) 10 collection information records
4) 10 STAC records
etc.

It does require that the client understands the granularity of the items which make up the collection.

**Comment by jratike80** [comment]
The client should be able to estimate easily if the request is within the limits so that it can generate good requests. The server should be able to send an immediate message "No way, I won't do that". Usually that should be simple but with re-projected subsets/bboxes, scaling, and when request goes partially beyound the coverage extent there is a little bit more to calculate.

**Comment by jerstlouis** [comment]
@cmheazel those numbers for different things likely correspond to very different processing and bandwidth resources, if we're talking about a multi-API resource (like from `{collectionId}` or landing page).

**Comment by jratike80** [comment]
As a background information you can have a look at the different request limits that can be configured for GeoServer. Trying to safeguard your server is not totally trivial https://docs.geoserver.org/master/en/user/services/wcs/configuration.html#request-limits

**Comment by jerstlouis** [comment]
And Features limits:

https://docs.geoserver.org/stable/en/user/services/wfs/webadmin.html

**Comment by cmheazel** [comment]
@jerstlouis agreed. The resources used to process the limit parameter are specific to the resource type being accessed.  If you are serving both Features and Coverages, that would be two different processes.  

If our goal is to make things easy for the client, then the limit should be intuitively obvious. We bundle Geospatial Data in collections. Those collections are a homogeneous (for now) set of items. The items in a collection should be delivered intact (what use is a partial Feature?). So the limit parameter should restrict the number of items to be returned from the requested collection.  Yes, a result set of 100 Features requires more bytes than a result set of 100 pixels, but the item count is the same.

**Comment by Schpidi** [comment]
see also https://github.com/opengeospatial/ogc_api_coverages/issues/54

**Comment by jerstlouis** [comment]
@cmheazel 

>  The items in a collection should be delivered intact (what use is a partial Feature?).

Well, _generalized_ features are useful as you can still render them (and may preserve the properties intact e.g. in the case of large but detailed polygonal features), which could be based on a maximum number of vertices resolution (similar to a coverage resolution). This works particularly well if your back-end is vector tiles.

For example our server currently automatically generalizes features when you ask for a large extent rather than doing paging, and we would like to communicate that the client will receive generalized features if a request would potentially result in too many vertices.
